apple/internal/aspects/docc_bundle_info_aspect.bzl

```diff
@@ -0,0 +1,92 @@
+"""Defines an aspect for collecting the DocCBundleInfo provider for targets that support it."""
+
+load(
+    "@build_bazel_rules_swift//swift/internal:providers.bzl",
+    "SwiftInfo",
+)
+load(
+    "//apple:providers.bzl",
+    "DocCBundleInfo",
+)
+
+_DOCC_BUNDLE_INFO_ASPECT_ATTRS = ["data", "deps"]
+
+def _top_level_swift_info(target, ctx):
+    """Returns the first SwiftInfo provider in the target or its deps"""
+    if SwiftInfo in target:
+        return target[SwiftInfo]
+    elif hasattr(ctx.rule.attr, "deps"):
+        for dep in ctx.rule.attr.deps:
+            if SwiftInfo in dep:
+                return dep[SwiftInfo]
+    return None
+
+def _swift_symbol_graph(target, swift_info):
+    """Returns the symbol graph from a SwiftInfo provider or fails if it doesn't exist."""
+    direct_modules = swift_info.direct_modules
+    if len(direct_modules) != 1:
+        fail("Expected target %s to have exactly one direct module in its SwiftInfo provider" % target.label)
+    module = direct_modules[0]
+    if not module.swift:
+        fail("Expected target %s to have a swift module in its SwiftInfo provider" % target.label)
+    swift_module = module.swift
+    if not swift_module.symbol_graph:
+        fail("Expected target %s to have a symbol graph in its SwiftInfo provider" % target.label)
+    return swift_module.symbol_graph
+
+def _top_level_docc_bundle(target, ctx):
+    """Returns the first .docc bundle for the target or its deps by looking in it's data."""
+
+    if hasattr(ctx.rule.attr, "data"):
```

Comment:
Same here, we should not traverse all target `data`, just for the packaging rules.